state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1215,154 +1209,241 @@ pub mod pallet {
 				request_id,
 			);
 
-			CcmSwapState::GasSwapScheduled { gas_swap_id: swap_id, principal_output_amount }
+			GasSwapState::Scheduled { gas_swap_id }
 		}
 
-		fn process_swap_outcomes(swaps: &[SwapState]) {
-			for swap in swaps {
-				let swap_request_id = swap.swap.swap_request_id;
+		fn refund_failed_swap(swap: Swap) {
+			let swap_request_id = swap.swap_request_id;
 
-				let Some(mut request) = SwapRequests::<T>::take(swap_request_id) else {
-					log_or_panic!("Swap request {swap_request_id} not found");
-					continue;
-				};
+			let Some(request) = SwapRequests::<T>::take(swap_request_id) else {
+				log_or_panic!("Swap request {swap_request_id} not found");
+				return;
+			};
 
-				let Some(output_amount) = swap.final_output else {
-					log_or_panic!("Swap {} is not completed yet!", swap.swap_id());
-					continue;
-				};
+			Self::deposit_event(Event::<T>::SwapRequestCompleted { swap_request_id: request.id });
 
-				Self::deposit_event(Event::<T>::SwapExecuted {
-					swap_request_id,
-					swap_id: swap.swap_id(),
-					// To be consistent with `swap_output` and `intermediate_amount` (which do
-					// not include the network fee), we report input amount without the network fee
-					// for swaps from STABLE_ASSET:
-					input_amount: if swap.input_asset() == STABLE_ASSET {
-						swap.stable_amount.unwrap_or_else(|| {
-							log_or_panic!("stable amount must be set for swaps from STABLE_ASSET");
-							swap.input_amount()
-						})
-					} else {
-						swap.input_amount()
-					},
-					input_asset: swap.input_asset(),
-					network_fee: swap.network_fee_taken.unwrap_or_default(),
-					output_asset: swap.output_asset(),
-					output_amount,
-					intermediate_amount: swap.intermediate_amount(),
-				});
+			let Some(refund_params) = request.refund_params else {
+				log_or_panic!("Trying to refund swap request {swap_request_id}, but missing refund parameters");
+				return;
+			};
 
-				match &mut request.state {
-					SwapRequestState::Ccm { state, ccm_deposit_metadata, output_address } =>
-						match state {
-							CcmSwapState::PrincipalSwapScheduled {
-								principal_swap_id,
-								gas_budget,
-								other_gas_asset,
-							} => {
-								debug_assert!(swap.swap_id() == *principal_swap_id);
-
-								if let Some(other_gas_asset) = other_gas_asset {
-									*state = Self::schedule_ccm_gas_swap(
-										swap_request_id,
-										request.input_asset,
-										*other_gas_asset,
-										*gas_budget,
-										output_amount,
-									);
+			match request.state {
+				SwapRequestState::Regular {
+					dca_state: DCAState { remaining_input_amount, accumulated_output_amount, .. },
+					output_address,
+				} => {
+					// Refund the failed swap and any unused input amount:
+					Self::egress_for_swap(
+						request.id,
+						swap.input_amount + remaining_input_amount,
```

Comment:
I don't understand this - why refund the `input_amount + remaining_input_amount` and not just the remaining amount?

Reply:
`remaining_input_amount` is the amount to be used in future chunks. When a chunk fails, we still want to refund it, just like we are doing for non-DCA swaps in 1.5. You could argue that `remaining_input_amount` should include the  amount from the currently scheduled swap, but I don't know if one approach is clearly better than the other. Regular swaps didn't have `remaining_input_amount`, so we were refunding `swap.input_amount`. The current approach is just a natural progression of that code.